File: src/main/java/com/example/reactive/chapter1/future/FutureShoppingCartService.java
package com.example.reactive.chapter1.future;

import java.util.concurrent.Future;
import java.util.concurrent.FutureTask;

import com.example.reactive.chapter1.commons.Input;
import com.example.reactive.chapter1.commons.Output;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@RequiredArgsConstructor
public class FutureShoppingCartService implements ShoppingCartService {

    @Override
    public Future<Output> calculate(Input input) {

        log.info("Future Style calculating...");

        FutureTask<Output> futureTask = new FutureTask<>(() -> {
            Thread.sleep(2000);
            log.info("Complete!");
            return new Output();
        });

        new Thread(futureTask).start();

        return futureTask;
    }
}

File: src/main/kotlin/com/example/reactor/store/future/FutureShoppingCartService.kt
package com.example.reactor.store.future

import com.example.reactor.shared.logger
import com.example.reactor.store.domain.Input
import com.example.reactor.store.domain.Output
import java.util.concurrent.Future
import java.util.concurrent.FutureTask

class FutureShoppingCartService {
    fun calculate(input: Input): Future<Output> {
        val future = FutureTask {
            Thread.sleep(2_000)
            logger().info("calculated!")
            Output()
        }
        Thread(future).start()
        return future
    }
}
